### PR TITLE
[SILGen] Fall back to module lookup on invalid conformance when building subst. maps for vtable thunks

### PR DESCRIPTION
**Pull request by AnthonyLatsis**
```swift
class Base<T> {
  func foo<U>(arg: U) where T == Int
}
class Derived<T>: Base<T> {
  func foo(<U>(arg: U) where T: Protocol
}
```
Here, as usual, we combine the override substitutions `T := Int` with the invocation generic signature `<T where T: Protocol>` to form a new substitution map for the reabstraction thunk, only to realize that we cannot recover the `Int : Protocol` conformance from just the original substitution map. As a consequence, we end up with an invalid conformance, and a crash later on once we actually emit the thunk. As ugly as it is, falling back to module lookup in `SubstitutionMap::get` fixes the issue.

<s>### FIX 2

Now that we allow purely contextual where clauses, we can face non-generic methods
with generic signatures that constrain all outer params to concrete types. E.g.
```swift
class Base<T> {
  func foo() where T == Int
}
class Derived<T>: Base<T> {
  func foo() where T: Protocol
}
```
Remove the assertion guarding against these generic signatures on SILFunctionType
so that we always retain the override substitutions needed to transform the derived
method's formal type when we emit a reabstraction thunk.</s>

### CODE REVIEW

**Comment by AnthonyLatsis** [comment]
cc @slavapestov and @jckarter

File: lib/AST/ASTContext.cpp [hunk 1]
```diff
@@ -3297,10 +3297,6 @@ SILFunctionType::SILFunctionType(
   }
         
   if (genericSig) {
-    assert(!genericSig->areAllParamsConcrete() &&
-           "If all generic parameters are concrete, SILFunctionType should "
-           "not have a generic signature at all");
-
```
Comment (AnthonyLatsis) [inline lib/AST/ASTContext.cpp]: Is there any way to assert that the original declaration is non-generic? It does not seem to be possible to recover it from here..

**Comment by slavapestov** [comment]
Can you try to come up with fixes that don't break these existing invariants?

**Comment by AnthonyLatsis** [comment]
I see, so we *do* lose performance.. I take it whatever eventually causes the performance regression does not check if the generic signature is at all useful.

> Can you try to come up with fixes that don't break these existing invariants?

I did try something, but hit another one saying ≈ `non-generic SIL function type not allowed to have a generic environment`. I am not well versed in this yet, so I would experiment a bit more in that direction if you think it is worth exploring as an alternative to your old PR.

> If you want to pursue the second fix further, you should finish off my old PR that removes the assertion: #23291
> 
> The hard part was that we lose performance because existing optimizations are currently disabled for
>  generics (such as inlining) but we still want them to take place in the fully-concrete case. So some 
> optimization passes will need changes to handle this.

 Optimization passes are interesting, but I'm hesitant about the idea of making the optimizer responsible for collecting unused generic signatures (unused because the function type is already fully substituted through canonicalization) just because we need that signature in another place within the SILGen library. Does this make sense to you?

**Comment by AnthonyLatsis** [comment]
I am going to rip off the second fix so that we can proceed with the first one. If it looks sane, that is. I have no idea how often the module lookup fallback is going to be hit in the wild though, and this *does* fix an edge case, so maybe we should test compiler performance?

**Comment by AnthonyLatsis** [comment]
FWIW, I refactored the first part into a (hopefully temporary) targeted fix to make sure we don't end up getting performance regressions elsewhere.

**Comment by jckarter** [comment]
It would be more correct if we could capture the conformance for `Int: Protocol` we saw at the point we establish the override. Falling back to conformance lookup in the module is a bug, since it's an opportunity to potentially end up with a different answer from the first time we established the conformance.

**Comment by AnthonyLatsis** [comment]
> Falling back to conformance lookup in the module is a bug, since it's an opportunity to potentially end up with a different answer from the first time we established the conformance.

Do you have an example of such a scenario? I'm curious.

**Comment by jckarter** [comment]
If `Protocol` is defined in a module that doesn't itself conform `Int` to `Protocol`, then two different modules could each declare their own retroactive conformance of `Int: Protocol`.

**Comment by slavapestov** [comment]
Yeah, I think @jckarter's answer is right. We need to record 'override substitutions' when we check the override, just like we record substitutions for protocol witnesses.

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Pull request closed without merging**